### PR TITLE
fix: changing keys from dictionary to object with two properties

### PR DESCRIPTION
**Pull request by ysalomon**

### CODE REVIEW

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 1]
```diff
@@ -427,8 +434,17 @@ Deceased <--+ Diagnosed +----> Recovered
                     <list style="hanging">
                         <t hangText="keys">
                             <vspace/>
-                            REQUIRED. A dictionary where the key is a URI of a key type and the value is an array
-                            of strings representing the diagnosis keys for that key type.
+                            REQUIRED. An object denoting the diagnosis keys. It has the following properties:
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Why not just make these both top level?
```
- key_type
- keys
```

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 2]
```diff
@@ -557,8 +573,18 @@ Deceased <--+ Diagnosed +----> Recovered
                     <list style="hanging">
                         <t hangText="keys">
                             <vspace/>
-                            REQUIRED. A dictionary where the key is an URI of a key type and the value is an array
-                            of strings representing the keys for that key type.
+                            REQUIRED. An object denoting the potential exposure keys. It has the following properties:
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Same as above

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 3]
```diff
@@ -807,8 +833,18 @@ Deceased <--+ Diagnosed +----> Recovered
                     <list style="hanging">
                         <t hangText="keys">
                             <vspace/>
-                            REQUIRED. A dictionary where the key is a URI of a key type and the value is an array
-                            of strings representing the diagnosis keys for that key type.
+                            REQUIRED. An object denoting the population of known diagnosis keys. It has the following
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Same as above

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 4]
```diff
@@ -853,8 +889,17 @@ Deceased <--+ Diagnosed +----> Recovered
                     <list style="hanging">
                         <t hangText="keys">
                             <vspace/>
-                            REQUIRED. A dictionary where the key is a URI of a key type and the value is an array
-                            of strings representing the diagnosis keys for that key type.
+                            REQUIRED. An object denoting the diagnosis keys. It has the following properties:
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Same as above

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 5]
```diff
@@ -853,8 +889,17 @@ Deceased <--+ Diagnosed +----> Recovered
...
+                            <list style="hanging">
+                                <t hangText="type">
+                                    <vspace/>
+                                    REQUIRED. A URI of a key type.
+                                </t>
+                                <t hangText="value">
+                                    <vspace/>
+                                    REQUIRED. an array of strings representing the diagnosis keys for that key type.
+                                </t>
+                            </list>
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: You need to update the examples in the appendix too!

**Review by sethyates — CHANGES_REQUESTED** [review]

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 6]
```diff
@@ -425,10 +432,13 @@ Deceased <--+ Diagnosed +----> Recovered
...
                     <list style="hanging">
+                        <t hangText="key_type">
+                            <vspace/>
+                            REQUIRED. A URI of a key type.
+                        </t>
                         <t hangText="keys">
                             <vspace/>
-                            REQUIRED. A dictionary where the key is a URI of a key type and the value is an array
-                            of strings representing the diagnosis keys for that key type.
+                            REQUIRED. an array of strings representing the diagnosis keys for that key type.
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Capitalise "An"

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 7]
```diff
@@ -555,10 +565,13 @@ Deceased <--+ Diagnosed +----> Recovered
...
                     <list style="hanging">
+                        <t hangText="key_type">
+                            <vspace/>
+                            REQUIRED. A URI of a key type.
+                        </t>
                         <t hangText="keys">
                             <vspace/>
-                            REQUIRED. A dictionary where the key is an URI of a key type and the value is an array
-                            of strings representing the keys for that key type.
+                            REQUIRED. an array of strings representing the potential exposure keys for that key type.
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Capitalise "An"

**Review by sethyates — CHANGES_REQUESTED** [review]

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 8]
```diff
@@ -1582,21 +1581,20 @@ Content-Type: application/json
...
-            "4075DB637A914896B5459EF3E377AF0D",
-            "5F51F91099B84C9685CD268941A3DEEC",
-            "67684C56F9A6407A99DB57E6B871EC1D",
-            "AA0817D30C64407CB4F5DE1425336922",
-            "CFC046DEF50244D58CA58A0AD7FED287",
-            "D555096A727C49FEADD03488154E6152",
-            "E48A37FB87924885AD6879857DC297B2"
-        ]
-    }
+    "key_type" : "https://tenprotocol.org/keys/btle",
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml:1631]: If we return `key_type` why not also return `threat`?

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 9]
```diff
@@ -1470,25 +1478,22 @@ Content-Type: application/json
...
-            "CFC046DEF50244D58CA58A0AD7FED287",
-            "D555096A727C49FEADD03488154E6152",
-            "E48A37FB87924885AD6879857DC297B2"
-        ]
-    },
-    "after": "2020-04-01T00:00:00",
-    "threats": [
-        "https://tenprotocol.org/threats/diseases/covid19"
+    "key_type" : "https://tenprotocol.org/keys/btle",
+    "keys": [
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml:1526]: We're returning `key_type` - should we return `threat` as well?

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 10]
```diff
@@ -1582,21 +1581,20 @@ Content-Type: application/json
...
+        "24BFE7A43563495BA36C79A96617A602",
+        "371108A73B964915A62D81E49A783F85",
+        "4075DB637A914896B5459EF3E377AF0D",
+        "5F51F91099B84C9685CD268941A3DEEC",
+        "67684C56F9A6407A99DB57E6B871EC1D",
+        "AA0817D30C64407CB4F5DE1425336922",
+        "CFC046DEF50244D58CA58A0AD7FED287",
+        "D555096A727C49FEADD03488154E6152",
+        "E48A37FB87924885AD6879857DC297B2"
+    ]
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: We might want to think of adding before/after to allow the implementation to page results and allow for cursoring.

**Review by sethyates — CHANGES_REQUESTED** [review]

File: draft-yates-threat-exposure-notification-protocol-00.xml (hunk 9 above)
Comment (ysalomon) [reply draft-yates-threat-exposure-notification-protocol-00.xml:1526]: I did think about it...

File: draft-yates-threat-exposure-notification-protocol-00.xml (hunk 10 above)
Comment (ysalomon) [reply draft-yates-threat-exposure-notification-protocol-00.xml]: Good call

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 11]
```diff
@@ -645,20 +661,33 @@ Deceased <--+ Diagnosed +----> Recovered
...
                         </t>
                         <t hangText="exposures">
                             <vspace/>
                             OPTIONAL. An array of days when exposures happened formatted as per
                             <xref target="RFC3339"/>
                             without a time component.
                         </t>
+                        <t hangText="before">
+                            <vspace/>
+                            REQUIRED. The time of the most recent upload across all diagnosis keys considered.
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Should note inclusive/exclusive.

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 12]
```diff
@@ -645,20 +661,33 @@ Deceased <--+ Diagnosed +----> Recovered
...
+                        </t>
+                        <t hangText="after">
+                            <vspace/>
+                            OPTIONAL. The time of the the oldest upload across all diagnosis keys considered.
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml:687]: Should note inclusive/exclusive.

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 13]
```diff
@@ -711,26 +740,25 @@ Deceased <--+ Diagnosed +----> Recovered
...
+                            REQUIRED. A URI of a key type.
                         </t>
-                        <t hangText="before">
+                        <t hangText="after">
                             <vspace/>
-                            OPTIONAL. A date to search for exposures before, formatted as specified
-                            in
-                            <xref target="RFC3339"/>
-                            with no time zone offset.
+                            OPTIONAL. A time to fetch for diagnosis keys uploaded at or after, formatted as specified
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Should note inclusive/exclusive.

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 14]
```diff
@@ -711,26 +740,25 @@ Deceased <--+ Diagnosed +----> Recovered
...
+                            in <xref target="RFC3339"/>
+                            with no time zone offset. If omitted, the server SHOULD provide
+                            a reasonable default based on the specified threats.
                         </t>
-                        <t hangText="after">
+                        <t hangText="before">
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml:758]: Should note inclusive/exclusive.

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 15]
```diff
@@ -805,10 +834,26 @@ Deceased <--+ Diagnosed +----> Recovered
...
+                        </t>
                         <t hangText="keys">
                             <vspace/>
-                            REQUIRED. A dictionary where the key is a URI of a key type and the value is an array
-                            of strings representing the diagnosis keys for that key type.
+                            REQUIRED. An array of URI's for the requested key types.
+                        </t>
+                        <t hangText="before">
+                            <vspace/>
+                            REQUIRED. The time of the most recent upload across all diagnosis keys considered.
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml]: Should note inclusive/exclusive.

File: draft-yates-threat-exposure-notification-protocol-00.xml [hunk 16]
```diff
@@ -805,10 +834,26 @@ Deceased <--+ Diagnosed +----> Recovered
...
+                        </t>
+                        <t hangText="after">
+                            <vspace/>
+                            OPTIONAL. The time of the the oldest upload across all diagnosis keys considered.
```
Comment (sethyates) [inline draft-yates-threat-exposure-notification-protocol-00.xml:857]: Should note inclusive/exclusive.

**Review by sethyates — CHANGES_REQUESTED** [review]

**Pull request merged**